drivers/gpio/gpio_bflb_bl60x_70x.c

```diff
@@ -0,0 +1,363 @@
+ */
+
+#define DT_DRV_COMPAT bflb_bl60x_70x_gpio
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/irq.h>
+#include <zephyr/dt-bindings/pinctrl/bflb-common-pinctrl.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#if defined(CONFIG_SOC_SERIES_BL60X)
+#include <zephyr/dt-bindings/pinctrl/bl60x-pinctrl.h>
+#include <bouffalolab/bl60x/bflb_soc.h>
+#include <bouffalolab/bl60x/glb_reg.h>
+#include <bouffalolab/bl60x/hbn_reg.h>
+#elif defined(CONFIG_SOC_SERIES_BL70X)
+#include <zephyr/dt-bindings/pinctrl/bl70x-pinctrl.h>
+#include <bouffalolab/bl70x/bflb_soc.h>
+#include <bouffalolab/bl70x/glb_reg.h>
+#include <bouffalolab/bl70x/hbn_reg.h>
+#else
+#error Unsupported platform
+#endif
+
+#include <zephyr/drivers/clock_control/clock_control_bflb_common.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_bl60x_bl70x);
+
+#define GPIO_BFLB_FUNCTION_GPIO 11
+/* Medium drive strength, 0 to 3 */
+#define GPIO_BFLB_DRIVE_STRENGTH 1
+
+#define GPIO_BFLB_TRIG_MODE_SYNC_LOW       0
+#define GPIO_BFLB_TRIG_MODE_SYNC_HIGH      1
+#define GPIO_BFLB_TRIG_MODE_SYNC_LEVEL     2
+
+#define GPIO_BFLB_PIN_INT_PER_REG    10
+#define GPIO_BFLB_PIN_INT_REG_SIZE   3
+#define GPIO_BFLB_PIN_INT_REG_MSK    GENMASK(2, 0)
+#define GPIO_BFLB_PIN_REG_SIZE_SHIFT 2
+
+/* this driver supports only 32 GPIO, which happens to be the maximum on BL6 and 7 serie.
+ */
+
+struct gpio_bflb_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	uint32_t base_reg;
+	void (*irq_config_func)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+};
+
+struct gpio_bflb_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+
+static int gpio_bflb_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+
+	*value = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL30_OFFSET);
+
+	return 0;
+}
+
+
+static int gpio_bflb_port_set_masked_raw(const struct device *dev,
+					 uint32_t mask,
+					 uint32_t value)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp = (tmp & ~mask) | (mask & value);
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp = tmp | mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static int gpio_bflb_port_clear_bits_raw(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp = tmp & ~mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
```

Comment:
This is a pattern that we are repeating too much so you could create functions like sys_set32 and sys_clear32 functions.

void sys_clear32(address, data)
{
	uint32_t tmp;

	tmp = sys_read32(address);
	tmp &= data;
	sys_write32(tmp, address);
}

	sys_clear32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET, mask);